csrc/python_frontend/python_bindings.cpp

```diff
@@ -171,6 +184,89 @@ std::vector<std::optional<bool>> computeContiguity(
+//   sizes   = [2, 1, 3, 1, 4, 3]
+//   strides = [12, 4, 4, 4, 1, 0]
+// we would compute stride order as: [5, 4, 3, 2, 1, 0]. Since the original
+// stride is in descending order. Note that there's more than one way to define
+// a stride order when we have equal strides. In the context of index
+// collapsing, how we resolve that shouldn't matter, hence we just go with
+// preserving their original order. Similarly, we compute contiguity as: [True,
+// None, True, None, True, None], Since the physical order is the same as the
+// logical order, this one is trivial to compute.
+//
+// e.g. Given size and stride as follow:
+//   sizes   = [2, 3, 1, 5, 4]
+//   strides = [28, 4, 14, 0, 1]
+// stride_order would be: [4, 2, 3, 0, 1], marking the order of strides in the
+// vector. Meanwhile, contiguity would be computed on the physical domain, i.e.
+// on sorted sizes & strides.
+//   sorted_size    = [2, 1, 3, 4, 5]
+//   sorted_strides = [28, 14, 4, 1, 0]
+//   contiguity would be: [False, None, True, True, None]
+//
+// This function returns a tuple of <contiguity, stride_order>
+std::tuple<std::vector<std::optional<bool>>, std::vector<int64_t>>
+computeTensorDescriptor(
+    const std::vector<int64_t>& sizes,
+    const std::vector<int64_t>& strides) {
+  NVF_CHECK(
+      sizes.size() == strides.size(),
+      "compute_contiguity: Sizes and strides must have the same number of dimensions");
+  std::vector<DimInfo> dim_info_vec;
+  for (auto i : c10::irange(sizes.size())) {
+    // NOTE: not supporting negative stride yet.
+    NVF_CHECK(strides[i] >= 0, "negative stride on tensor is not supported");
+    dim_info_vec.emplace_back(DimInfo{(int64_t)i, sizes[i], strides[i]});
+  }
+  // sort by stride
+  std::sort(
+      dim_info_vec.begin(),
+      dim_info_vec.end(),
+      [](const auto& l, const auto& r) { return l.stride > r.stride; });
+
+  // Dimensions are marked contiguous by inspecting the current dimension and
+  // one to the right towards the inner dimension while skipping over broadcast
+  // dimensions.
+  // The innermost dimension, that is not broadcasted, does not have any
+  // dimension to it's right and needs to have stride equal to 1 in order to be
+  // marked contiguous.
+  for (int64_t i = 0; i < (int64_t)sizes.size();) {
+    dim_info_vec[i].stride_order = (int64_t)sizes.size() - 1 - i;
+    if (dim_info_vec[i].notBroadcast()) {
+      auto l = i++;
+      int64_t expected = 1;
+      for (; i < (int64_t)sizes.size(); i++) {
+        dim_info_vec[i].stride_order = (int64_t)sizes.size() - 1 - i;
+        if (dim_info_vec[i].notBroadcast()) {
+          expected = dim_info_vec[i].stride * dim_info_vec[i].size;
+          break;
+        }
+      }
+      dim_info_vec[l].contiguity = (dim_info_vec[l].stride == expected);
+    } else {
+      i++;
+    }
+  }
+
+  std::vector<std::optional<bool>> contiguity(sizes.size(), std::nullopt);
+  std::vector<int64_t> stride_order(sizes.size(), -1);
```

Comment:
This would probably be clearer in two loops so that setting stride order doesn't need to be repeated.

Reply:
> so that setting stride order doesn't need to be repeated.

I wasn't quite following the second part. but I will split the for loop. :stuck_out_tongue_winking_eye: